src/action/finder/download.rs

```diff
@@ -0,0 +1,295 @@
+use crate::action::finder::chunks::{
+    get_file_chunks, Chunks, GetFileChunksConfig,
+};
+use crate::action::finder::request::{
+    DownloadActionOptions, HashActionOptions,
+};
+use crate::fs::Entry;
+use rrg_proto::file_finder_download_action_options::OversizedFilePolicy as DownloadOversizedFilePolicy;
+use rrg_proto::file_finder_hash_action_options::OversizedFilePolicy as HashOversizedFilePolicy;
+use sha2::{Digest, Sha256};
+use std::fs::File;
+use std::io::{BufReader, Take};
+
+#[derive(Debug)]
+pub enum Response {
+    /// Download action is not performed and no further action is required.
+    Skip(),
+    /// File was not downloaded, but hash action must be executed.
+    HashRequest(HashActionOptions),
+    /// Chunks of data to be uploaded.
+    CollectData(Chunks<BufReader<Take<File>>>),
+}
+
+/// Performs `download` action logic and returns file contents to be uploaded
+/// or another action to be executed.
+pub fn download(entry: &Entry, config: &DownloadActionOptions) -> Response {
+    if entry.metadata.len() > config.max_size {
+        match config.oversized_file_policy {
+            DownloadOversizedFilePolicy::Skip => {
+                return Response::Skip();
+            }
+            DownloadOversizedFilePolicy::DownloadTruncated => (),
+            DownloadOversizedFilePolicy::HashTruncated => {
+                let hash_config = HashActionOptions {
+                    max_size: config.max_size,
+                    oversized_file_policy:
+                        HashOversizedFilePolicy::HashTruncated,
+                };
+                return Response::HashRequest(hash_config);
+            }
+        };
+    }
+
+    let chunks = get_file_chunks(
+        &entry.path,
+        &GetFileChunksConfig {
+            max_read_bytes: config.max_size,
+            bytes_per_chunk: config.chunk_size,
+            start_offset: 0,
+            overlap_bytes: 0,
+        },
+    );
+
+    match chunks {
+        Some(chunks) => Response::CollectData(chunks),
+        None => Response::Skip(),
+    }
+}
+
+/// A type representing unique identifier of a given chunk.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct ChunkId {
+    /// A SHA-256 digest of the referenced chunk data.
+    sha256: [u8; 32],
+    offset: u64,
+    length: u64,
+}
+
+impl ChunkId {
+    /// Creates a chunk identifier for the given chunk.
+    pub fn make(chunk: &Vec<u8>, offset: u64) -> ChunkId {
```

Comment:
Prefer to use slices rather than references to vectors, it is almost always better.

Reply:
Changed to slice.